### PR TITLE
Enhancement/auto scaling

### PR DESCRIPTION
**Pull request by MickVermeulen**
This creates server auto scaling where inactive servers are automatically shut down given a set of parameters. Health is checked but a number of heart beats per interval (both specified in the settings), whenever a world server sends enough heart beats to the master server it will consider them healthy. I've been testing this on the test server for quite some time and it seems to be working quite smoothly.

### CODE REVIEW

**Comment by TheNexusAvenger** [comment]
I mentioned in private messages about dumping information about the world, like the logs and zone objects, on world close in case it helps leads to finding the problem for the 55% loading bug. Is this going to happen?

**Comment by MickVermeulen** [comment]
I looked into it but this generally shuts down worlds before 55% would even occur as worlds are shut down after a timeframe of inactivity. It's currently not possible to programmatically retrieve information about 55% afaik (otherwise we'd probably have fixed it by now :P).

**Review by TheNexusAvenger — APPROVED** [review]

File: Uchu.World/WorldUchuServer.cs [hunk 1]
```diff
@@ -5,6 +5,7 @@
 using System.Net;
 using System.Reflection;
 using System.Threading.Tasks;
+using System.Timers;
```
Comment (enteryournamehere) [inline Uchu.World/WorldUchuServer.cs]: System.Timers isn't used in this file

File: Uchu.World/Objects/Zone.cs [hunk 2]
```diff
@@ -13,6 +13,7 @@
 using RakDotNet;
 using RakDotNet.IO;
 using Sentry;
+using Uchu.Api.Models;
```
Comment (enteryournamehere) [inline Uchu.World/Objects/Zone.cs]: Uchu.Api.Models isn't used in this file

File: Uchu.Master/MasterServer.cs [hunk 3]
```diff
@@ -82,6 +89,53 @@ private static async Task Main(string[] args)
...
+                            var health when health >= 0.5 => ServerHealth.SeverelyLagging,
+                            var health when health >= 0.25 => ServerHealth.Unhealthy,
+                            
+                            // If hardly any heartbeats were received, gradually downgrade the health until closing
+                            _ => (ServerHealth) (int) InstanceHealth[id] - 1
+                        };
+                    }
+                    else
+                    {
+                        InstanceHealth.Add(id, ServerHealth.Dead);
```
Comment (enteryournamehere) [inline Uchu.Master/MasterServer.cs:115]: If an instance doesn't have a known health yet it's set to Dead here, but in the next `foreach` all Dead instances are killed. I might be missing something obvious here but why wouldn't all new instances be killed as soon as this check runs?

**Review by enteryournamehere — COMMENTED** [review]
2 minor comments and one question, besides that it all looks good to me!

File: Uchu.Master/MasterServer.cs (hunk 3 above)
Comment (MickVermeulen) [reply Uchu.Master/MasterServer.cs:115]: This is just the case for a world that exists but for which the health is unknown, this does not actually create a new world, it's indeed intended to automatically kill a world for which the status is not known. When a new world is actually instanced (see MasterServer.cs, line 385), its status is set to healthy.

File: Uchu.World/WorldUchuServer.cs (hunk 1 above)
Comment (MickVermeulen) [reply Uchu.World/WorldUchuServer.cs]: Good call.

File: Uchu.World/Objects/Zone.cs (hunk 2 above)
Comment (MickVermeulen) [reply Uchu.World/Objects/Zone.cs]: Good call.

File: Uchu.Master/MasterServer.cs (hunk 3 above)
Comment (enteryournamehere) [reply Uchu.Master/MasterServer.cs:115]: That clears it up, thanks!

**Review by enteryournamehere — APPROVED** [review]

**Pull request merged**